kernel/src/vm/mod.rs

```diff
@@ -49,3 +65,225 @@ pub fn mem_total() -> usize {
+where
+    F: FnMut(&Range<Vaddr>, &Frame<dyn AnyUFrameMeta>, &PageProperty) -> Result<(), ()>,
+{
+    // Copy range.start -> range.start + page.
+    let search_end = start + PROMOTED_PAGE_SIZE;
+    cursor.jump(start).map_err(|_| ())?;
+
+    while cursor.virt_addr() < search_end {
+        // Query under the current cursor. If the virtual address is unmapped then we
+        // fail and cannot map a huge page.
+        let (sub_range, sub_mapping) = cursor.query().map_err(|_| ())?;
+
+        // The mapping might not exist if the page hasn't been faulted in yet. We can
+        // skip over such mappings.
+        if let Some((ref sub_frame, sub_props)) = sub_mapping {
+            cb(&sub_range, sub_frame, &sub_props)?;
+        }
+
+        // Advance the cursor to the end of this mapping
+        cursor.jump(sub_range.end).map_err(|_| ())?;
+    }
+
+    Ok(())
+}
+
+fn promote_hugepages(proc: &Arc<Process>, addr_hint: Option<Vaddr>) -> Result<(), ()> {
+    // Ensure that the current process doesn't run until we have scanned it's mappings
+    let _ = PauseProcGaurd::new(proc.clone());
+
+    let proc_vm = proc.vm();
+    let proc_vm_guard = proc_vm.lock_root_vmar();
+    let proc_vmar = proc_vm_guard.unwrap();
+    let preempt_guard = disable_preempt();
+    let space_len = proc_vmar.size();
+    let mut cursor = match proc_vmar
+        .vm_space()
+        .cursor_mut(&preempt_guard, &(0..space_len))
+    {
+        Ok(cursor) => cursor,
+        _ => {
+            return Ok(());
+        }
+    };
+
+    // If we have an address hint, jump the cursor to that address, and only consider a single
+    // region for promotion.
+    if let Some(addr_hint) = addr_hint {
+        cursor.jump(addr_hint).map_err(|_| ())?;
+        space_len = PROMOTED_PAGE_SIZE;
+    }
+
+    while cursor.find_next(space_len - cursor.virt_addr()).is_some() {
+        let (range, _) = match cursor.query() {
+            Ok(v) => v,
+            Err(_) => break,
+        };
+
+        // If the address is not hugepage aligned go to the next mapping
+        if range.start % PROMOTED_PAGE_SIZE != 0 {
+            let next = range.start - range.start % PROMOTED_PAGE_SIZE + PROMOTED_PAGE_SIZE;
+            if next < space_len {
+                if let Err(_) = cursor.jump(next) {
+                    break;
+                }
+            } else {
+                break;
+            }
+            continue;
+        }
+
+        if (range.end - range.start) >= PROMOTED_PAGE_SIZE {
+            // Already huge, nothing to do here
+            continue;
+        }
+
+        let start = range.start;
+
+        let mut props: Option<PageProperty> = None;
+        // Track if any sub pages were accessed or dirty
+        let mut accessed = false;
+        let mut dirty = false;
+
+        // Find all subpages in this region and copy them to the new frame.
+        let mut should_remap = true;
```

Comment:
Document this variable a bit. Because it's set in multiple places it's not clear what value it takes and why.